### PR TITLE
Kernel: Abstract platform-specific current time methods from Scheduler

### PR DESCRIPTION
**Pull request by supercomputer7**
This change ensures that the scheduler doesn't depend on a platform specific or arch-specific code when it initializes itself, but rather we ensure that in compile-time we will generate the appropriate code to find the correct arch-specific current time methods.

### CODE REVIEW

File: Kernel/Arch/CurrentTime.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Types.h>
+
+namespace Kernel {
+
+typedef u64 (*fptr)();
```
Comment (kleinesfilmroellchen) [inline Kernel/Arch/CurrentTime.h:13]: Can this be done more idiomatically, i.e
 via Function and using?

File: Kernel/Arch/CurrentTime.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+fptr optional_arch_specific_current_time();
```
Comment (kleinesfilmroellchen) [inline Kernel/Arch/CurrentTime.h:15]: Isn't `optional_current_time` enough?

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Kernel/Arch/CurrentTime.h (hunk 2 above)
Comment (supercomputer7) [reply Kernel/Arch/CurrentTime.h:15]: Yes it is, I will change it :)

File: Kernel/Arch/CurrentTime.h (hunk 1 above)
Comment (supercomputer7) [reply Kernel/Arch/CurrentTime.h:13]: Unfortunately no, I tried that and the compiler insists that __dso_handle is not implemented when I try to change things to work with the `Function` container, due to a required change in the `Scheduler.cpp` file:
https://github.com/SerenityOS/serenity/blob/0c07bed89c58d3fca745fbe12753335247406c05/Kernel/Scheduler.cpp#L53-L55

When I change that line to be `Function<u64(void)> Scheduler::current_time;` it doesn't work with the __dso_handle error.

Comment (supercomputer7) [reply Kernel/Arch/CurrentTime.h:13]: Turns out it is too much hacky and ugly, so I will leave it this way.

**Comment by ADKaster** [comment]
I merged that TSC-must-be-stable PR that caused this one to have conflicts, sorry about that!

**Comment by supercomputer7** [comment]
> I merged that TSC-must-be-stable PR that caused this one to have conflicts, sorry about that!

This is fixed now, so hopefully this will pass CI and can be merged then.

File: Kernel/Arch/x86/CurrentTime.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+{
+    return read_tsc();
+}
+
+fptr optional_current_time()
+{
+    VERIFY(Processor::is_initialized()); // sanity check
+    // Figure out a good scheduling time source
+    if (Processor::current().has_feature(CPUFeature::TSC) && Processor::current().has_feature(CPUFeature::CONSTANT_TSC)) {
+        // TODO: only use if TSC is running at a constant frequency?
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/Arch/x86/CurrentTime.cpp]: Done.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Pull request merged**